File: src/main/java/io/github/willslaunwhite/linkedlist/problems/SwapNodes.java
package io.github.willslaunwhite.linkedlist.problems;

import io.github.willslaunwhite.linkedlist.ListNode;

public class SwapNodes {
    public ListNode swapNodes(ListNode head, int k) {
        ListNode left = head, right = head;
        while (k > 1) {
            right = right.next;
            k--;
        }
        ListNode dummy = right;
        while (dummy.next != null) {
            left = left.next;
            dummy = dummy.next;
        }
        int temp = right.val;
        right.val = left.val;
        left.val = temp;

        return head;

    }
}

File: python/linked_list/problems/swap_nodes.py
from typing import Optional

from linked_list.list_node import ListNode


def swap_nodes(head: Optional[ListNode], k: int) -> Optional[ListNode]:
    left = right = head
    while k > 1 and right:
        right: Optional[ListNode] = right.next
        k -= 1
    dummy: Optional[ListNode] = right
    while dummy and dummy.next and left:
        left: Optional[ListNode] = left.next
        dummy = dummy.next
    if right and left:
        temp: int = right.val
        right.val = left.val
        left.val = temp
    return head
